From ea8709b7ef92f5ab8d87a68f9b6977e6c34c0c04 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Marek=20Marczykowski-G=C3=B3recki?=
 <marmarek@invisiblethingslab.com>
Date: Sat, 23 May 2015 04:25:01 +0200
Subject: [PATCH] Add 'nostrictreset' attribute to PCI host devices
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This allows to assign PCI device to some VM, even when the device does not
support any reset method. This may be dangerous in some cases (especially when
the device is later assigned to some other VM). But in some cases it still
makes sense - for example when the device is assigned to the same VM whole the
time.

Signed-off-by: Marek Marczykowski-Górecki <marmarek@invisiblethingslab.com>
---
 docs/formatdomain.rst             |  2 ++
 src/conf/domain_conf.c            | 12 ++++++++++++
 src/conf/domain_conf.h            |  1 +
 src/conf/schemas/domaincommon.rng |  5 +++++
 src/hypervisor/virhostdev.c       |  2 ++
 src/util/virpci.c                 | 18 ++++++++++++++++++
 src/util/virpci.h                 |  2 ++
 7 files changed, 42 insertions(+)

diff --git a/docs/formatdomain.rst b/docs/formatdomain.rst
index 9420921712..de974b4350 100644
--- a/docs/formatdomain.rst
+++ b/docs/formatdomain.rst
@@ -4892,6 +4892,8 @@ or:
       to the guest. This framebuffer allows the vgpu to be used as a boot display
       before the gpu driver is loaded within the guest. ``ramfb`` requires the
       ``display`` attribute to be set to ``on``.
+      Additionally when ``strictreset`` is ``no``, device will be assigned to
+      the domain, even when reset fails. The default is ``yes``.
    ``scsi``
       For SCSI devices, user is responsible to make sure the device is not used
       by host.
diff --git a/src/conf/domain_conf.c b/src/conf/domain_conf.c
index 379cbdc37c..fb12ddbc36 100644
--- a/src/conf/domain_conf.c
+++ b/src/conf/domain_conf.c
@@ -6555,6 +6555,7 @@ virDomainHostdevDefParseXMLSubsys(xmlNodePtr node,
     virDomainHostdevSubsysSCSIVHost *scsihostsrc = &def->source.subsys.u.scsi_host;
     virDomainHostdevSubsysMediatedDev *mdevsrc = &def->source.subsys.u.mdev;
     virTristateBool managed;
+    g_autofree char *nostrictreset = NULL;
     g_autofree char *model = NULL;
     virDomainDeviceSGIO sgio;
     virTristateBool rawio;
@@ -6569,6 +6570,11 @@ virDomainHostdevDefParseXMLSubsys(xmlNodePtr node,
                                         VIR_XML_PROP_NONE, &managed));
     virTristateBoolToBool(managed, &def->managed);
 
+    if ((nostrictreset = virXMLPropString(node, "nostrictreset")) != NULL) {
+        if (STREQ(nostrictreset, "yes"))
+            def->nostrictreset = true;
+    }
+
     model = virXMLPropString(node, "model");
 
     /* @type is passed in from the caller rather than read from the
@@ -25772,6 +25778,8 @@ virDomainActualNetDefFormat(virBuffer *buf,
         virDomainHostdevDef *hostdef = virDomainNetGetActualHostdev(def);
         if  (hostdef && hostdef->managed)
             virBufferAddLit(buf, " managed='yes'");
+        if  (hostdef && hostdef->nostrictreset)
+            virBufferAddLit(buf, " nostrictreset='yes'");
     }
     if (def->trustGuestRxFilters)
         virBufferAsprintf(buf, " trustGuestRxFilters='%s'",
@@ -26068,6 +26076,8 @@ virDomainNetDefFormat(virBuffer *buf,
     virBufferAsprintf(buf, "<interface type='%s'", typeStr);
     if (hostdef && hostdef->managed)
         virBufferAddLit(buf, " managed='yes'");
+    if (hostdef && hostdef->nostrictreset)
+        virBufferAddLit(buf, " nostrictreset='yes'");
     if (def->trustGuestRxFilters)
         virBufferAsprintf(buf, " trustGuestRxFilters='%s'",
                           virTristateBoolTypeToString(def->trustGuestRxFilters));
@@ -28338,6 +28348,8 @@ virDomainHostdevDefFormat(virBuffer *buf,
     if (def->mode == VIR_DOMAIN_HOSTDEV_MODE_SUBSYS) {
         virBufferAsprintf(buf, " managed='%s'",
                           def->managed ? "yes" : "no");
+        if (def->nostrictreset)
+            virBufferAddLit(buf, " nostrictreset='yes'");
 
         if (def->source.subsys.type == VIR_DOMAIN_HOSTDEV_SUBSYS_TYPE_SCSI &&
             scsisrc->sgio)
diff --git a/src/conf/domain_conf.h b/src/conf/domain_conf.h
index 59fb00abd0..77f5c28aea 100644
--- a/src/conf/domain_conf.h
+++ b/src/conf/domain_conf.h
@@ -374,6 +374,7 @@ struct _virDomainHostdevDef {
     bool readonly;
     bool shareable;
     virTristateBool writeFiltering;
+    bool nostrictreset;
     union {
         virDomainHostdevSubsys subsys;
         virDomainHostdevCaps caps;
diff --git a/src/conf/schemas/domaincommon.rng b/src/conf/schemas/domaincommon.rng
index 6347d80597..94b02b5508 100644
--- a/src/conf/schemas/domaincommon.rng
+++ b/src/conf/schemas/domaincommon.rng
@@ -3851,6 +3851,11 @@
               <ref name="virYesNo"/>
             </attribute>
           </optional>
+          <optional>
+            <attribute name="nostrictreset">
+              <ref name="virYesNo"/>
+            </attribute>
+          </optional>
           <interleave>
             <element name="source">
               <optional>
diff --git a/src/hypervisor/virhostdev.c b/src/hypervisor/virhostdev.c
index 720ed9bb60..27c96bd4d7 100644
--- a/src/hypervisor/virhostdev.c
+++ b/src/hypervisor/virhostdev.c
@@ -257,6 +257,8 @@ virHostdevGetPCIHostDevice(const virDomainHostdevDef *hostdev,
         return -1;
     }
 
+    virPCIDeviceSetStrictReset(actual, !hostdev->nostrictreset);
+
     *pci = g_steal_pointer(&actual);
     return 0;
 }
diff --git a/src/util/virpci.c b/src/util/virpci.c
index bdf81e8839..34fc70e261 100644
--- a/src/util/virpci.c
+++ b/src/util/virpci.c
@@ -91,6 +91,7 @@ struct _virPCIDevice {
     /**/
 
     bool          managed;
+    bool          strictreset;
 
     virPCIStubDriver stubDriverType;
     char            *stubDriverName; /* if blank, use default for type */
@@ -1246,6 +1247,9 @@ virPCIDeviceReset(virPCIDevice *dev,
                        dev->name,
                        err ? err->message :
                        _("no FLR, PM reset or bus reset available"));
+        if (!dev->strictreset)
+            /* do not fail */
+            ret = 0;
     }
 
  cleanup:
@@ -1982,6 +1986,8 @@ virPCIDeviceNew(const virPCIDeviceAddress *address)
 
     virPCIDeviceAddressCopy(&dev->address, address);
 
+    dev->strictreset    = true;
+
     dev->name = virPCIDeviceAddressAsString(&dev->address);
 
     dev->path = g_strdup_printf(PCI_SYSFS "devices/%s/config", dev->name);
@@ -2096,6 +2102,18 @@ virPCIDeviceGetManaged(virPCIDevice *dev)
     return dev->managed;
 }
 
+void
+virPCIDeviceSetStrictReset(virPCIDevice *dev, bool strictreset)
+{
+    dev->strictreset = strictreset;
+}
+
+unsigned int
+virPCIDeviceGetStrictReset(virPCIDevice *dev)
+{
+    return dev->strictreset;
+}
+
 void
 virPCIDeviceSetStubDriverType(virPCIDevice *dev, virPCIStubDriver driverType)
 {
diff --git a/src/util/virpci.h b/src/util/virpci.h
index 933099da6c..fd9c08b181 100644
--- a/src/util/virpci.h
+++ b/src/util/virpci.h
@@ -135,6 +135,8 @@ virPCIStubDriver virPCIDeviceGetStubDriverType(virPCIDevice *dev);
 void virPCIDeviceSetStubDriverName(virPCIDevice *dev,
                                    const char *driverName);
 const char *virPCIDeviceGetStubDriverName(virPCIDevice *dev);
+void virPCIDeviceSetStrictReset(virPCIDevice *dev, bool strictreset);
+unsigned int virPCIDeviceGetStrictReset(virPCIDevice *dev);
 virPCIDeviceAddress *virPCIDeviceGetAddress(virPCIDevice *dev);
 void virPCIDeviceSetUsedBy(virPCIDevice *dev,
                            const char *drv_name,
-- 
2.54.0

